README.md
⚠️ Se vuoi rimanere aggiornato, ti puoi iscrivere al [feed RSS dei ChangeLog](https://github.com/opendatasicilia/covid-open-report-sicilia/commits/main/CHANGELOG.md.atom)

# COVID open report Sicilia
[![Frictionless](https://github.com/opendatasicilia/covid-open-report-sicilia/actions/workflows/frictionless.yaml/badge.svg)](https://repository.frictionlessdata.io/report?user=opendatasicilia&repo=covid-open-report-sicilia&flow=frictionless) [![License: CC BY 4.0](https://img.shields.io/badge/License-CC%20BY%204.0-lightgrey.svg)](https://creativecommons.org/licenses/by/4.0/) <a href="https://www.datibenecomune.it/"><img src="https://img.shields.io/badge/%F0%9F%99%8F-%23datiBeneComune-%23cc3232"/></a>

Dal 6 ottobre 2021 il Dipartimento Attività Sanitarie e Osservatorio Epidemiologico (DASOE) della Regione Siciliana rilascia bollettini settimanali (in formato PDF) ai quali sono allegate due tabelle contenenti **dati epidemiologici e vaccinali con dettaglio comunale**. Le tabelle allegate vengono convertite in CSV e caricate nella cartella [`dati`](https://github.com/opendatasicilia/covid-open-report-sicilia/tree/main/dati) di questo repo.

Gli stessi dati alimentano:
- 📍 API endpoints documentati e raggiungibili al [seguente link](https://covid-open-report-sicilia.herokuapp.com/);
- 📊 dashboard interattiva raggiungibile al [seguente link](https://report-dasoe.opendatasicilia.it/).

![covid-open-report-sicilia](https://user-images.githubusercontent.com/77018886/143773850-a0f79d60-77e3-4a8c-bb82-5553c85c0bcf.png)

## Bollettini pubblicati
- [Report 06 Ottobre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-10/Report%20Completo%2006%20Ottobre%202021.pdf)
- [Report 13 Ottobre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-10/Report%20Completo%2013%20Ottobre%202021.pdf)
- [Report 20 Ottobre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-10/Report%20Completo%2020%20Ottobre%202021.pdf)
- [Report 27 Ottobre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-10/Report%20Completo%2027%20Ottobre%202021.pdf)
- [Report 04 Novembre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-11/Bollettino%20n%C3%82%C2%B0%205%20del%204%20novembre.pdf)
- [Report 10 Novembre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-11/Bollettino%20settimanale%20n%C2%B06%20del%2010%20novembre%202021.pdf)
- [Report 17 Novembre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-11/Bollettino%20Dasoe%207%20del%2017%20Novembre%202021.pdf)
- [Report 24 Novembre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-11/Bollettino%20n.8%20del%2024%20novembre%202021.pdf)
- [Report 01 Dicembre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-12/Bollettino%20settimanale%201%20dicembre%20%282%29.pdf)
- [Report 08 Dicembre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-12/Bollettino%2008%20Dicembre%202021.pdf)
- [Report 15 Dicembre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-12/Bollettino%20Dasoe%20n.11%20del%2015%20Dicembre%202021.pdf)
- [Report 22 Dicembre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-12/Bollettino%20Dasoe%2012%20del%2022%20Dicembre%202021.pdf)
- [Report 29 Dicembre 2021.pdf](https://www.regione.sicilia.it/sites/default/files/2021-12/Bollettino%20Dasoe%2013%20del%2029%20dicembre%202021.pdf)
- [Report 06 Gennaio 2022.pdf](https://www.regione.sicilia.it/sites/default/files/2022-01/bollettino%2014%20finale.pdf)
- [Report 12 Gennaio 2022.pdf](https://www.regione.sicilia.it/sites/default/files/2022-01/Bollettino%2012%20gennaio%202022.pdf)
- [Report 19 Gennaio 2022.pdf](https://www.regione.sicilia.it/sites/default/files/2022-01/Bollettino%2016%20del%2019%20Gennaio%202022.pdf)

## Struttura repository
```
covid-open-report-sicilia
├── 📂dati
│   ├── 📂incidenza
│   │   ├── 📄incidenza-YYYYMMDD.csv
│   │   ├── 📄incidenza-latest.csv
│   │   └── 📄incidenza.csv
│   └── 📂vaccini
│       ├── 📄vaccini-YYYYMMDD.csv
│       ├── 📄vaccini-latest.csv
│       └── 📄vaccini.csv
├── 📂download
│   ├── 📄report.csv
│   └── 📄report-YYYYMMDD.pdf
├── 📂script
│   ├── 📄script.sh
│   └── 📄script.py
└── 📂api
```

## Schema dati
- metadata: [datapackage.yaml](https://github.com/opendatasicilia/covid-open-report-sicilia/blob/main/datapackage.yaml)
### Dati per comune relativi a incidenza e nuovi casi settimanali

- Directory:  dati/incidenza<br>
- Struttura file settimanale: `incidenza-YYYYMMDD.csv`<br>
- File complessivo: `incidenza.csv`<br>
- File ultimi dati (latest): `incidenza-latest.csv`
- Encoding: `UTF-8`
- Separatore di campo: `,`

Campo | Descrizione | Formato | Esempio
-- | -- | -- | --
data | Data pubblicazione report DASOE | YYYY-MM-DD | 2021-10-27
cod_prov | Codice ISTAT della Provincia | Numero | 84
pro_com_t | Codice ISTAT del Comune | Testo | 084002
provincia | Denominazione della Provincia | Testo | Agrigento
comune | Denominazione del Comune | Testo | Alessandria della Rocca
incidenza | Incidenza cumulativa settimanale (ogni 100.000 abitanti) | Numero | 855
casi | Nuovi casi settimanali | Numero | 6

### Dati per comune relativi a persone vaccinate

- Directory:  dati/vaccini<br>
- Struttura file settimanale: `vaccini-YYYYMMDD.csv`<br>
- File complessivo: `vaccini.csv`<br>
- File ultimi dati (latest): `vaccini-latest.csv`
- Encoding: `UTF-8`
- Separatore di campo: `,`

Campo | Descrizione | Formato | Esempio
-- | -- | -- | --
data | Data pubblicazione report DASOE | YYYY-MM-DD | 2021-10-27
cod_prov | Codice ISTAT della Provincia | Numero | 84
pro_com_t | Codice ISTAT del Comune | Testo | 084002
provincia | Denominazione della Provincia | Testo | Agrigento
comune | Denominazione del Comune | Testo | Alessandria della Rocca
prima_dose | Percentuale di persone vaccinate con almeno una dose (calcolata rispetto al target) | Numero | 84.65
seconda_dose | Percentuale di: persone vaccinate con 2 o più dosi, persone vaccinate in monodose per pregressa infezione Covid, persone vaccinate con Janssen (calcolata rispetto al target) | Numero | 82.72

**Nota bene:**
- Dal [Bollettino n°4](https://www.regione.sicilia.it/sites/default/files/2021-11/Bollettino%20n%C3%82%C2%B0%205%20del%204%20novembre.pdf) si apprende che _Per target si intende la popolazione residente ISTAT 2021 DI Età >= 12 ANNI._
- Dal [Bollettino n°12](https://www.regione.sicilia.it/sites/default/files/2021-12/Bollettino%20Dasoe%2012%20del%2022%20Dicembre%202021.pdf) si apprende che a partire dal [Bollettino n° 11 del 15/12/2021](https://www.regione.sicilia.it/sites/default/files/2021-12/Bollettino%20Dasoe%20n.11%20del%2015%20Dicembre%202021.pdf) gli allegati relativi alla vaccinazione contemplano _il target 5-11 anni quale platea avente diritto alla vaccinazione, con un conseguente decremento della % di popolazione immunizzata per singolo comune._ Pertanto è possibile stimare il numero assoluto di persone vaccinate utilizzando [questo file json.](https://raw.githubusercontent.com/opendatasicilia/cors-dashboard/main/src/data/targets.json)

## Licenza
<a href="https://creativecommons.org/licenses/by/4.0/"><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/1/16/CC-BY_icon.svg/640px-CC-BY_icon.svg.png" width="150"/></a>

## Elaborazioni
Dati aperti di questo tipo consentono la realizzazione di svariate elaborazioni:
- [Report settimanale interattivo](https://report-dasoe.opendatasicilia.it/) di [Open Data Sicilia](https://opendatasicilia.it) ([GitHub](https://github.com/opendatasicilia/cors-dashboard))
- [Sintesi Bollettino settimanale: Dati Epidemiologici e Vaccinali](https://opendatasicilia.github.io/OpenDataSicilia-per-il-Coronavirus/vaccini/report_sintesi/) di [Giovan Battista Vitrano](https://twitter.com/gbvitrano)
- [MONIVAX - Monitora i vaccinati della tua città](https://github.com/opendatasicilia/monivax) di [Giovanni Pirrotta](https://twitter.com/gpirrotta)
- [Mappa - variazione percentuale di nuovi casi](https://gjrichter.github.io/viz/COVID-19/gallery/ODS%20-%20Report/) di [Guenter Richter](https://twitter.com/grichter?s=09)


CHANGELOG.md
# CHANGELOG

In questo file verranno indicate tutte le modifiche apportate al presente repository per data di rilascio.

## 2022-01-17
Modifica: I campi `%vaccinati` e `%immunizzati` sono stati rinominati rispettivamente in `prima_dose` e `seconda_dose`. I file soggetti alle modifiche sono quelli di seguito riportati:
- datapackage.yaml
- README.md
- api/routes/vaccini.js
- data/vaccini/*.csv
- scripts/convert.py


scripts/convert.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import pandas as pd
from PyPDF2 import PdfFileReader
import tabula

path = '.'
comuni = pd.DataFrame(pd.read_csv('https://raw.githubusercontent.com/gabacode/vaxExtract/main/utilities/Elenco-comuni-siciliani.csv', converters={'pro_com_t': '{:0>6}'.format}))
latest = pd.DataFrame(pd.read_csv(path+'/download/report.csv')).iloc[-1]
file = latest['nome_file']
date = latest['data_report']


def getRanges(file):
    '''
    Estrae i range delle pagine
    Allegato 1 e Allegato 2
    '''
    result_list = []
    r_pages = []

    reader = PdfFileReader(file,strict=False)
    pages = reader.numPages

    for page_number in range(0, pages):
        page = reader.getPage(page_number)
        page_content = page.extractText()

        if "ALLEGATO" in page_content:
            result = {"page": page_number}
            result_list.append(result)

    r_pages.append(pages)

    for i in range(len(result_list)):
        r_pages.append(result_list[i]['page'])

    r_pages.sort()

    allegati = {
        "file": file,
        "incidenza": [r_pages[0], r_pages[1]],
        "vaccini": [r_pages[1]+1, r_pages[2]]
    }

    return allegati


def isDigit(x):
    '''
    True se x contiene un numero
    '''
    try:
        float(x)
        return True
    except ValueError:
        return False


def getVax(vax):
    '''
    Estrae tabella Vaccini da file PDF
    '''
    # Leggi il PDF  VAX con tabula-py
    print('Leggo tabella Vaccini...attendi...')
    pages = list(range(vax['vaccini'][0], vax['vaccini'][1]+1))
    pdf = tabula.read_pdf(vax['file'], pages=pages, pandas_options={'header': None}, multiple_tables=True, columns=['103, 607'])
    print('Ho letto.')

    # Unisci in un unico dataframe e bonifica i dati
    vax = pd.concat(pdf).reset_index(drop=True)
    vax = vax.dropna(thresh=3).dropna(subset=[0])
    vax = vax[~vax[0].str.contains("Totale", na=False)]

    for index, row in vax.iterrows():
        if(row[0] == "Santo Stefano di"):
            row[0] = "Santo Stefano di Camastra"
            
    vax = vax.drop([1,2,3], axis=1)
    vax.columns = ['comune', 'totale']
    vax[['prima_dose','seconda_dose']] = vax.totale.str.split(expand=True)
    vax = vax.drop(['totale'], axis=1)
    
    vax['prima_dose'] = vax['prima_dose'].str.replace(',', '.').str.rstrip('%')
    vax['seconda_dose'] = vax['seconda_dose'].str.replace(',', '.').str.rstrip('%')
    
    vax.reset_index(drop=True, inplace=True)

    # Carica l'helper comuni siciliani
    out = pd.merge(vax, comuni, on='comune', how='inner')
    out = out[['cod_prov', 'pro_com_t', 'provincia','comune', 'prima_dose', 'seconda_dose']]
    out['prima_dose'] = out['prima_dose'].str.replace(',', '.').str.rstrip('%')
    out['seconda_dose'] = out['seconda_dose'].str.replace(',', '.').str.rstrip('%')
    out.insert(0, 'data', date)

    # Controlla che ci siano tutti i comuni
    assert (len(out) == 390 or 391), "Errore: Sono presenti meno comuni del previsto."
   
    # Esporta CSV
    lastUpdate = open(path+'/dati/vaccini/vaccini.csv').read().rsplit('\n', 2)[1].split(',')[0]
    if date != lastUpdate:
        print('Esporto CSV...')
        out.to_csv(path+'/dati/vaccini/vaccini-'+date.replace("-", "")+'.csv', index=None, header=True)
        out.to_csv(path+'/dati/vaccini/vaccini-latest.csv', index=None, header=True)
        out.to_csv(path+'/dati/vaccini/vaccini.csv', mode='a', index=None, header=False)
    csv = path+'/dati/vaccini/vaccini-'+date.replace("-", "")+'.csv'

    return csv


def getIncidenza(pdf):
    '''
    Estrae l'incidenza da file PDF
    '''
    # Legge le pagine relative all'incidenza
    print('Leggo tabella Incidenza...attendi...')
    reader = PdfFileReader(pdf['file'],strict=False)
    pages = pdf['incidenza']
    
    # Looppa le pagine, rimuovi numero della pagina e sostituisce breaklines
    # aggiunge ad un'unica, grande stringa
    textes = []
    try:
        for i in range(pages[0], pages[-1]+1):
            page = reader.getPage(i)
            text = page.extractText()
            text = text.replace('\n', ' ')
            textes.append(text[2::])
    except Exception as e:
        print(e)

    # Seleziona parti di interesse, formatta elementi ambigui e ritorna una lista
    out = ' '.join(textes)\
        .rpartition('settimane')[2]\
        .rpartition('Totale')[0]\
        .replace('- ', '-')\
        .replace('---', '0%')\
        .replace('  ', ' ')\
        .replace('  ', ' ')\
        .replace('  ', ' ')\
        .replace('%', '')\
        .replace("O'", "Ò").replace("I'", "Ì").replace("U'", "Ù")\
        .split()
    
    # Riconosce il nome del comune rispetto ad un numero e ritorna una nuova lista
    new = ""
    for split in out:
        if not isDigit(split):
            new = new + split + ' '
        if isDigit(split):
            new = new + ',' + split + ','
    new = new.replace(',,', ',').replace(
        ' ,', ',').replace(' -', '-').split(',')

    # Looppa la lista e crea tuple a gruppi di 4
    it = iter(new)
    data = list(zip(it, it, it, it))
    
    # Crea dataframe
    df = pd.DataFrame(data, columns=['comune', 'casi', 'incidenza', 'variazione'])
    df = df[['comune', 'incidenza', 'casi']]
    
    # Rimuovi distretti (duplicati)
    incidenza = df[~df["comune"].duplicated(keep="last")]
    incidenza.reset_index(inplace=True, drop=True)

    for index, row in incidenza.iterrows():
        if(row[0] == "SANTA CATERINA"):
            row[0] = "Santa Caterina Villarmosa"
    
    # Inner join e recupera info comuni
    out = pd.merge(incidenza, comuni, left_on=incidenza["comune"].str.lower(), right_on=comuni["comune"].str.lower(), how="inner")
    out.rename(columns={'comune_y': 'comune'}, inplace=True)
    out = out[['cod_prov', 'pro_com_t', 'provincia', 'comune', 'incidenza', 'casi']].sort_values(by=['provincia', 'comune'])
    out.reset_index(drop=True, inplace=True)
    out.insert(0, 'data', date)

    # Controlla che ci siano tutti i comuni
    assert (len(out) == 390 or 391), "Errore: Sono presenti meno comuni del previsto."

    # Esporta CSV
    lastUpdate = open(path+'/dati/incidenza/incidenza.csv').read().rsplit('\n', 2)[1].split(',')[0]
    if date != lastUpdate:
        print('Esporto CSV...')
        out.to_csv(path+'/dati/incidenza/incidenza-'+date.replace("-", "")+'.csv', index=None, header=True)
        out.to_csv(path+'/dati/incidenza/incidenza-latest.csv', index=None, header=True)
        out.to_csv(path+'/dati/incidenza/incidenza.csv', mode='a', index=None, header=False)
    csv = path+'/dati/incidenza/incidenza-'+date.replace("-", "")+'.csv'

    return csv

def addToReadme():
    '''
    Aggiunge ultima riga del report.csv al file README.md
    '''
    mesi = {"01":"Gennaio","02":"Febbraio","03":"Marzo","04":"Aprile","05":"Maggio","06":"Giugno","07":"Luglio","08":"Agosto","09":"Settembre","10":"Ottobre","11":"Novembre","12":"Dicembre"}
    data = date.split('-')
    data = data[2] + " " + mesi[data[1]] + " " + data[0]

    title_index = ""
    with open("./README.md", "r+", encoding="utf-8") as f:
        lines = f.readlines()
        for index, line in enumerate(lines):
            if "Bollettini pubblicati" in line:
                title_index = index
        insert_index = title_index + int(latest['n'])
        try:
            lastDay = lines[insert_index].split(' ')[2]
        except:
            lastDay = lines[insert_index-1].split(' ')[2]
        newDay = date.split('-')[2]
        if lastDay != newDay:
            insert_content = "- [Report " + data + ".pdf](" + latest['URL'] + ")\n"
            lines.insert(insert_index, insert_content)
            f.seek(0)
            f.writelines(lines)
        f.seek(0)
    f.close()

def getAbs(vax):
    vaccini = pd.read_csv(vax, converters={'pro_com_t': '{:0>6}'.format})
    target = pd.read_csv('https://raw.githubusercontent.com/opendatasicilia/comuni-italiani/main/dati/target5.csv', converters={'pro_com_t': '{:0>6}'.format})
    target = target[['pro_com_t','>=5']]
    target.columns = ['pro_com_t', 'target']
    out_abs = pd.merge(vaccini, target, on='pro_com_t', how='inner')
    out_abs['prima_dose'] = (out_abs['target'] * out_abs['prima_dose'].astype(float) / 100).round().astype(int)
    out_abs['seconda_dose'] = (out_abs['target'] * out_abs['seconda_dose'].astype(float) / 100).round().astype(int)
    out_abs.columns = ['data', 'cod_prov', 'pro_com_t', 'provincia', 'comune', 'prima_dose', 'seconda_dose', 'target']
    out_abs.to_csv(path+'/dati/vaccini/vaccini-abs-'+date.replace("-", "")+'.csv', index=None, header=True)
    out_abs.to_csv(path+'/dati/vaccini/vaccini-abs-latest.csv', index=None, header=True)

getIncidenza(getRanges(path+'/download/'+latest['nome_file']))
getVax(getRanges(path+'/download/'+latest['nome_file']))
addToReadme()

#getAbs(getVax(getRanges(path+'/download/'+latest['nome_file'])))
